File: box-and-box/ports/rust/src/reflexive.rs
// reflexive.rs — Reflexive Arithmetic / AGM policy revision (faithful port of reflexive.mjs).
// Laws R1–R8, RB1–RB3. The capstone guarantee is ENTRENCHMENT: a constitutional core is
// immutable to weakening.

use crate::norm::{Modality, Norm};
use crate::supervise::TemporalSpec;
use std::collections::HashSet;

pub struct Policy {
    pub norms: Vec<Norm>,
    pub specs: Vec<TemporalSpec>,
    pub entrenched: HashSet<String>,
}

impl Policy {
    pub fn new() -> Policy {
        Policy {
            norms: Vec::new(),
            specs: Vec::new(),
            entrenched: HashSet::new(),
        }
    }
    pub fn with_norms(norms: Vec<Norm>) -> Policy {
        Policy {
            norms,
            specs: Vec::new(),
            entrenched: HashSet::new(),
        }
    }
}

// amendments — the three legal moves over a body of norms
#[derive(Clone)]
pub enum Amendment {
    Enact { item: Item, authority: String, time: f64 },
    Repeal { id: String, authority: String, time: f64 },
    Amend { id: String, item: Item, authority: String, time: f64 },
}

#[derive(Clone)]
pub enum Item {
    Norm(Norm),
    Spec(TemporalSpec),
}

impl Item {
    #[allow(dead_code)]
    pub fn id(&self) -> &str {
        match self {
            Item::Norm(n) => &n.id,
            Item::Spec(s) => &s.id,
        }
    }
}

pub fn enact(item: Item) -> Amendment {
    Amendment::Enact { item, authority: "self".into(), time: 0.0 }
}
pub fn enact_at(item: Item, time: f64) -> Amendment {
    Amendment::Enact { item, authority: "self".into(), time }
}
pub fn repeal(id: &str) -> Amendment {
    Amendment::Repeal { id: id.to_string(), authority: "self".into(), time: 0.0 }
}
pub fn amend(id: &str, item: Item) -> Amendment {
    Amendment::Amend { id: id.to_string(), item, authority: "self".into(), time: 0.0 }
}

fn find_norm<'a>(policy: &'a Policy, id: &str) -> Option<&'a Norm> {
    policy.norms.iter().find(|n| n.id == id)
}

fn conflicts(a: &Norm, b: &Norm) -> bool {
    a.target.is_some()
        && a.target == b.target
        && ((a.modality == Modality::Obligatory && b.modality == Modality::Forbidden)
            || (a.modality == Modality::Forbidden && b.modality == Modality::Obligatory))
}

fn dedupe(arr: Vec<Norm>) -> Vec<Norm> {
    // last occurrence wins, preserving final order of last-seen keys (mirrors JS Map)
    let mut order: Vec<String> = Vec::new();
    let mut map: std::collections::HashMap<String, Norm> = std::collections::HashMap::new();
    for x in arr {
        if !map.contains_key(&x.id) {
            order.push(x.id.clone());
        }
        map.insert(x.id.clone(), x);
    }
    order.into_iter().map(|k| map.remove(&k).unwrap()).collect()
}

pub struct Admissible {
    pub ok: bool,
    pub reason: Option<String>,
}

// the reflexive guard: admissible only if it does not WEAKEN the entrenched core
pub fn admissible(policy: &Policy, am: &Amendment) -> Admissible {
    match am {
        Amendment::Repeal { id, .. } => {
            if policy.entrenched.contains(id) {
                Admissible { ok: false, reason: Some(format!("\u{201c}{}\u{201d} is entrenched — cannot be repealed", id)) }
            } else {
                Admissible { ok: true, reason: None }
            }
        }
        Amendment::Amend { id, item, .. } => {
            if !policy.entrenched.contains(id) {
                return Admissible { ok: true, reason: None };
            }
            let cur = find_norm(policy, id);
            let next = match item {
                Item::Norm(n) => Some(n),
                Item::Spec(_) => None,
            };
            match (cur, next) {
                (Some(cur), Some(next)) => {
                    let stronger = next.modality == cur.modality && next.priority >= cur.priority;
                    if stronger {
                        Admissible { ok: true, reason: None }
                    } else {
                        Admissible { ok: false, reason: Some(format!("amendment would weaken entrenched \u{201c}{}\u{201d}", id)) }
                    }
                }
                _ => Admissible { ok: false, reason: Some(format!("\u{201c}{}\u{201d} is entrenched — cannot be amended", id)) },
            }
        }
        Amendment::Enact { item, .. } => {
            if let Item::Norm(n) = item {
                for id in &policy.entrenched {
                    if let Some(e) = find_norm(policy, id) {
                        if conflicts(e, n) && n.priority >= e.priority {
                            return Admissible {
                                ok: false,
                                reason: Some(format!("enacted norm would override entrenched \u{201c}{}\u{201d}", id)),
                            };
                        }
                    }
                }
            }
            Admissible { ok: true, reason: None }
        }
    }
}

pub struct Arbitration {
    pub norms: Vec<Norm>,
    pub overridden: Vec<String>,
}

// arbitrate same-target conflicts: lex superior (priority) then lex posterior (recency)
pub fn arbitrate(norms: &[Norm]) -> Arbitration {
    let mut overridden: Vec<String> = Vec::new();
    for a in norms {
        for b in norms {
            if std::ptr::eq(a, b) || !conflicts(a, b) {
                continue;
            }
            let a_wins = a.priority > b.priority
                || (a.priority == b.priority && a.time > b.time);
            if a_wins && !overridden.contains(&b.id) {
                overridden.push(b.id.clone());
            }
        }
    }
    Arbitration {
        norms: norms
            .iter()
            .filter(|n| !overridden.contains(&n.id))
            .cloned()
            .collect(),
        overridden,
    }
}

pub struct Revision {
    pub policy: Policy,
    pub accepted: bool,
    pub overridden: Vec<String>,
}

fn clone_policy(p: &Policy) -> Policy {
    Policy {
        norms: p.norms.clone(),
        specs: p.specs.iter().map(clone_spec).collect(),
        entrenched: p.entrenched.clone(),
    }
}

fn clone_spec(s: &TemporalSpec) -> TemporalSpec {
    TemporalSpec {
        id: s.id.clone(),
        formula: s.formula.clone(),
        kind: s.kind,
        ctd: s.ctd.clone(),
    }
}

// revise the policy by an amendment, if admissible
pub fn revise(policy: &Policy, am: &Amendment) -> Revision {
    let adm = admissible(policy, am);
    if !adm.ok {
        return Revision {
            policy: clone_policy(policy),
            accepted: false,
            overridden: Vec::new(),
        };
    }
    let mut next = clone_policy(policy);
    match am {
        Amendment::Enact { item, authority, time } => match item {
            Item::Norm(n) => {
                let mut nn = n.clone();
                nn.time = *time;
                let _ = authority;
                next.norms.push(nn);
            }
            Item::Spec(s) => next.specs.push(clone_spec(s)),
        },
        Amendment::Repeal { id, .. } => {
            next.norms.retain(|n| &n.id != id);
            next.specs.retain(|s| &s.id != id);
        }
        Amendment::Amend { id, item, time, .. } => match item {
            Item::Norm(n) => {
                next.norms = next
                    .norms
                    .into_iter()
                    .map(|x| {
                        if &x.id == id {
                            let mut nn = n.clone();
                            nn.time = *time;
                            nn
                        } else {
                            x
                        }
                    })
                    .collect();
            }
            Item::Spec(s) => {
                next.specs = next
                    .specs
                    .into_iter()
                    .map(|x| if &x.id == id { clone_spec(s) } else { x })
                    .collect();
            }
        },
    }
    next.norms = dedupe(next.norms);
    let arb = arbitrate(&next.norms);
    next.norms = arb.norms;
    Revision {
        policy: next,
        accepted: true,
        overridden: arb.overridden,
    }
}

// entrenching is monotone
pub fn entrench(policy: &Policy, id: &str) -> Policy {
    let mut next = clone_policy(policy);
    next.entrenched.insert(id.to_string());
    next
}

pub fn policy_key(p: &Policy) -> String {
    let mut n: Vec<String> = p
        .norms
        .iter()
        .map(|x| format!("{}|{:?}|{}", x.id, x.modality, x.priority))
        .collect();
    n.sort();
    let mut s: Vec<String> = p.specs.iter().map(|x| x.id.clone()).collect();
    s.sort();
    let mut e: Vec<String> = p.entrenched.iter().cloned().collect();
    e.sort();
    format!("n={:?};s={:?};e={:?}", n, s, e)
}

pub struct Stabilization {
    pub policy: Policy,
    pub rounds: usize,
    pub stable: bool,
}

// reflective stability: apply proposals until the policy stops changing (a fixed point)
pub fn stabilize(policy: &Policy, proposals: &[Amendment], max_rounds: usize) -> Stabilization {
    let mut cur = clone_policy(policy);
    for round in 0..max_rounds {
        let mut changed = false;
        for am in proposals {
            let r = revise(&cur, am);
            if r.accepted && policy_key(&r.policy) != policy_key(&cur) {
                cur = r.policy;
                changed = true;
            }
        }
        if !changed {
            return Stabilization {
                policy: cur,
                rounds: round + 1,
                stable: true,
            };
        }
    }
    Stabilization {
        policy: cur,
        rounds: max_rounds,
        stable: false,
    }
}

File: witness/src/AmpersandBoxDesign/box-and-box/reflexive.mjs
// reflexive.mjs — Reflexive Arithmetic, faithful runtime (v0.5)
// The fifth rung: norms that revise themselves. The object being revised is a Policy —
// the deontic norms + temporal specs of rungs 3–4. Revision follows AGM belief-revision
// discipline (success, consistency, minimal change) and the deontic norm-change principles
// (lex superior = priority, lex posterior = recency). The capstone guarantee is ENTRENCHMENT:
// a constitutional core is immutable to weakening — the system can make itself MORE
// constrained, never less, so self-modification can never relax the safety floor. Laws R1–R8.
import { STATUS } from './norm.mjs';

export const Policy = (p = {}) => ({
  norms: [...(p.norms || [])],
  specs: [...(p.specs || [])],
  entrenched: new Set(p.entrenched || []) // ids of the constitutional core
});

// amendments — the three legal moves over a body of norms (enact / repeal / amend)
export const enact = (item, { authority = 'self', time = 0 } = {}) => ({ op: 'enact', item, authority, time });
export const repeal = (id, { authority = 'self', time = 0 } = {}) => ({ op: 'repeal', id, authority, time });
export const amend = (id, item, { authority = 'self', time = 0 } = {}) => ({ op: 'amend', id, item, authority, time });

const find = (policy, id) => policy.norms.find((n) => n.id === id) || policy.specs.find((s) => s.id === id);
const isNorm = (x) => !!x && x.modality !== undefined;
const conflicts = (a, b) => isNorm(a) && isNorm(b) && a.target != null && a.target === b.target &&
  ((a.modality === 'obligatory' && b.modality === 'forbidden') || (a.modality === 'forbidden' && b.modality === 'obligatory'));
const dedupe = (arr) => { const seen = new Map(); for (const x of arr) seen.set(x.id, x); return [...seen.values()]; };

// the reflexive guard: admissible only if it does not WEAKEN the entrenched core
export function admissible(policy, am) {
  if (am.op === 'repeal')
    return policy.entrenched.has(am.id) ? { ok: false, reason: `“${am.id}” is entrenched — cannot be repealed` } : { ok: true };
  if (am.op === 'amend') {
    if (!policy.entrenched.has(am.id)) return { ok: true };
    const cur = find(policy, am.id), next = am.item;
    if (!cur || !isNorm(cur)) return { ok: false, reason: `“${am.id}” is entrenched — cannot be amended` };
    const stronger = next.modality === cur.modality && (next.priority ?? 0) >= (cur.priority ?? 0); // strengthening allowed
    return stronger ? { ok: true } : { ok: false, reason: `amendment would weaken entrenched “${am.id}”` };
  }
  if (am.op === 'enact') {
    if (isNorm(am.item)) for (const id of policy.entrenched) {
      const e = find(policy, id);
      if (e && conflicts(e, am.item) && (am.item.priority ?? 0) >= (e.priority ?? 0))
        return { ok: false, reason: `enacted norm would override entrenched “${id}”` };
    }
    return { ok: true };
  }
  return { ok: false, reason: 'unknown op' };
}

// arbitrate same-target conflicts: lex superior (priority) then lex posterior (recency)
export function arbitrate(norms) {
  const overridden = [];
  for (const a of norms) for (const b of norms) {
    if (a === b || !conflicts(a, b)) continue;
    const aWins = (a.priority ?? 0) > (b.priority ?? 0) ||
      ((a.priority ?? 0) === (b.priority ?? 0) && (a.time ?? 0) > (b.time ?? 0));
    if (aWins && !overridden.includes(b.id)) overridden.push(b.id);
  }
  return { norms: norms.filter((n) => !overridden.includes(n.id)), overridden };
}

// the core operation: revise the policy by an amendment, if admissible
export function revise(policy, am) {
  const adm = admissible(policy, am);
  if (!adm.ok) return { policy, accepted: false, reason: adm.reason, changed: null, overridden: [] };
  const next = Policy(policy);
  next.entrenched = new Set(policy.entrenched);
  const stamp = (x) => ({ ...x, time: am.time ?? 0, authority: am.authority });
  if (am.op === 'enact') { if (isNorm(am.item)) next.norms.push(stamp(am.item)); else next.specs.push(stamp(am.item)); }
  else if (am.op === 'repeal') { next.norms = next.norms.filter((n) => n.id !== am.id); next.specs = next.specs.filter((s) => s.id !== am.id); }
  else if (am.op === 'amend') { next.norms = next.norms.map((n) => n.id === am.id ? stamp(am.item) : n); next.specs = next.specs.map((s) => s.id === am.id ? stamp(am.item) : s); }
  next.norms = dedupe(next.norms); next.specs = dedupe(next.specs); // enacting an in-force id refreshes, never duplicates
  const arb = arbitrate(next.norms);
  next.norms = arb.norms;
  return { policy: next, accepted: true, reason: `${am.op} “${am.item ? am.item.id : am.id}” accepted`, changed: am.op, overridden: arb.overridden };
}

// entrenching is monotone — you may add to the constitution, never (at this level) remove from it
export function entrench(policy, id) { const next = Policy(policy); next.entrenched = new Set(policy.entrenched); next.entrenched.add(id); return next; }

export const policyKey = (p) => JSON.stringify({
  n: p.norms.map((x) => [x.id, x.modality, x.priority ?? 0]).sort(),
  s: p.specs.map((x) => x.id).sort(), e: [...p.entrenched].sort()
});

// reflective stability: apply proposals until the policy stops changing (a fixed point)
export function stabilize(policy, proposals, { maxRounds = 12 } = {}) {
  let cur = policy; const log = [];
  for (let round = 0; round < maxRounds; round++) {
    let changed = false;
    for (const am of proposals) {
      const r = revise(cur, am);
      log.push({ round, op: am.op, accepted: r.accepted, reason: r.reason });
      if (r.accepted && policyKey(r.policy) !== policyKey(cur)) { cur = r.policy; changed = true; }
    }
    if (!changed) return { policy: cur, rounds: round + 1, stable: true, log };
  }
  return { policy: cur, rounds: maxRounds, stable: false, log };
}

export const digest = (p) => `${p.norms.length} norms · ${p.specs.length} specs · ${p.entrenched.size} entrenched`;

export default { Policy, enact, repeal, amend, admissible, arbitrate, revise, entrench, stabilize, policyKey, digest };
